Make Faraday-APRS Persistent Across Network Failures

kb1lqc opened the pull request:
# BASE91 APRS Comment Telemetry
[aprs.fi](www.aprs.fi) and a few other services support BASE91 telemetry transmission. This allows up to 13 bits of ADC telemetry per channel. This PR pulled in what was #269 so here is the  information relevant to that PR. @reillyeon beat @el-iso to the punch on there. I appreciate the help from both!

## Changelog
- Imports aprslib and uses base91 functionality. Please note that this is note a standard BASE91 implementation and is a custom APRS version of base91. Qeue [XKCD](https://xkcd.com/927/).
- Removes use of `sendTelemetry` function as it is no longer needed
- Only sends scaling values on first transmission to reduce unnecessary packet transmissions. This should be revisited later but seems good for now.
- Sends a `telemetrySequence` number that is unbounded to `sendPosition` as BASE91 comment telemetry can support up to 16 bits of counting and this should be fine for the forceable future :)
- `sendPosition` implements [BASE91 Comment Telemetry](http://he.fi/doc/aprs-base91-comment-telemetry.txt) with ADC0, ADC1, ADC3, ADC6, BOARDTEMP, and GPIO bits per #266.

Raw packets which end in BASE91 comment telemetry 
![image](https://user-images.githubusercontent.com/331540/30197057-2b826f0a-941a-11e7-95a8-b29ba9a6155e.png)

12 bit ADC data looks very nice!
![image](https://user-images.githubusercontent.com/331540/30197084-65efa518-941a-11e7-9e7a-331c38e1ad4e.png)

# Persistent APRS-IS Connection Across Network

This PR is based on #269 as it was branched from that PR repository and aims to improve reliability when connecting to APRS-IS servers across a network. As we've seen, often a few minutes to a few hours after starting APRS logging a network event will cause  socket ERRNO 10054 or similar and `faraday-aprs` could not recover. This was annoying at best and at worst rendered the program not terribly useful.

Mind helping me out @reillyeon, if no time I can see if @kb1lqd has some free time in the near future :)

## Changelog
- Created a `sendAPRSPacket` function to consolidate sending of data to the socket and exception catching
- Reconnects after a IOERROR in 2 second loop attempts

## Issue Tickets
This PR addresses issues related to the following tickets Faraday-Software #119, Faraday-Software #190, Faraday-Software #199, Faraday-Software #201, Faraday-Software #262. A few of these tickets can likely be completely closed as a result of this testing and eventual PR!

kb1lqc committed:
Initial addition of APRS BASE91 Code

Using aprslib with a simple import of base91 function to help create
position comments that will send highly accurate data (up to 16 bits).
Therefor this will help with using all 12 bits of the ADC on Faraday!.

kb1lqc committed:
ADC0, ADC1, ADC3, ADC6, and BOARDTEMP BASE91

all channels previously sent with telemetry packet are now encoded into
BASE91 format. They are however not used yet.

kb1lqc committed:
Added telemetry sequence to sendPosition

Needed it for BASE91 APRS transmissions. Also commented out sendTelemetry.

kb1lqc committed:
Working telemetry and some hacked in telemetry seq

I had to change up how the telemetry sequence counter worked. Also, we
actually can count higher than 255 with BASE91 telemetry so we will do
that. Also, this has a lot of development code but I will clean it up
later as I let it run for a while.

kb1lqc committed:
Per @rossengeorgiev there is a width for BASE91

He pointed out that I missed the width setting in his implementation of
BASE91. It appears to be working! My bad...

kb1lqc committed:
Cleaned up APRS Position Telemetry Code

Removed a bunch of dev code and added normal comments back in.

kb1lqc committed:
Added IO to position comment data

using integers encoded into BASE91 format I now also send IO data to
APRS-IS with each position!

kb1lqc committed:
Only send Parameters, Labels, and Equations once

Decided to only send them once right now during initial loop through
sending. This will hopefully reduce the number of times I have APRS.fi
tell me I'm abusing the network :)

kb1lqc committed:
Fixed up pytest PEP8 Failures

kb1lqc committed:
Updated comments for sendPosition() function

kb1lqc committed:
Reversed 4095 hardcoded values in sendTelemetry

I changed them for development testing and never changed them back. This
function is currently unused as it is.

kb1lqc committed:
Just got reconnection working!

Lots of development code in here but I can close down the aprs-is server
(aprsc) and start it back up after indication of an error by faraday-aprs.
The program just keeps on going after server starts back up.

kb1lqc committed:
Removed unnecessary comment in sendPositions

kb1lqc committed:
Simplified packet sending with sendAPRSPacket

Created a function that could be commong to all packet types. This
prevents me from having to duplicate the same code a bunch in aprs.py

kb1lqc committed:
Removed commented out code and added documentation

Removed old commented out code while also adding documentation for
sendAPRSPacket.

kb1lqc committed:
Send all packets on 10th loop

While still commented out, the labels, parameters, and equations packets
are now send on every 10th loop.

kb1lqc committed:
Implemented sendAPRSPacket() on all packets

Went through all packets being sent and made them all use the
sendAPRSPacket function. This helps clarify the design as well as provide
fault detection and reconnection.

kb1lqc committed:
Updated sendAPRSPacket documenation

Fixed some errors in the docstring.

kb1lqc committed:
Fixed pytest errors/warnings

kb1lqc committed:
Updated parameter names

ADC names were wrong and I made them not wrong.

kb1lqc committed:
Fixed reconnection issue on extended network loss

Running tests over the day I observed from a local aprsc server that
network connectivity was lost for a few minutes at the same time that my
APRS telemetry stopped. Investigating by disableing my own network
connection for several faraday-aprs loops I observed the ERRNO 10054,
ERRNO 11004, and ERRNO 9 errors. I found that while I closed the socket
aprsSock in the connection code on an IOError I did not recreate a new
socket for aprsSock which resulted in an indefinitely bad file descriptor
(because it didn't exist) whenever I tried to reconnect. This should be
fixed now as I added the automatic aprsSock creation in.

reillyeon commented on faraday/aprs.py:
```diff
@@ -103,6 +104,11 @@ def aprs_worker(config, sock):
     telemSequence = 0
 
+    #dev
+    sock = ''
```
This line is unnecessary.

kb1lqc replied:
Removed, `conn=False` now used in the initialization just above to force it to initialize on first loop. Thanks for catching this!

reillyeon commented on faraday/aprs.py:
```diff
@@ -114,11 +120,21 @@ def aprs_worker(config, sock):
+
+            # Just send labels, Parameters, and Equations every 10th loop
+            if telemSequence % 10 == 0:
+
```
nit: Extra blank line.

kb1lqc replied:
Removed

reillyeon commented on faraday/aprs.py:
```diff
@@ -103,6 +104,11 @@ def aprs_worker(config, sock):
     telemSequence = 0
 
+    #dev
+    sock = ''
```
nit: This statement is unnecessary.

kb1lqc replied:
?

reillyeon commented on faraday/aprs.py:
```diff
@@ -317,11 +382,8 @@ def sendPositions(stations, socket):
-                except IOError as e:
-                    logger.error(e)
+                status = sendAPRSPacket(socket, positionString)
+                return status
```
`return sendAPRSPacket(socket, positionString)`

Here and below.

kb1lqc replied:
All `return status` lines replaced with obviously better `return sendAPRSPacket()`. Thanks for observing this unnecessary addition of variables!

reillyeon commented on faraday/aprs.py:
```diff
@@ -693,14 +714,17 @@ def connectAPRSIS():
             else:
                 logger.info("Connection successful!")
                 return aprssock
                 break
```
Not this change, but `break` after `return` does nothing.

kb1lqc replied:
Removed

reillyeon commented on faraday/aprs.py:
```diff
@@ -762,7 +786,8 @@ def main():
     logger.info('Starting Faraday APRS-IS application')
-    sock = connectAPRSIS()
+    #sock = connectAPRSIS()
+    sock = ''
```
Why pass `sock` to `aprs_worker` at all if it calls `connectAPRSIS()` itself?

kb1lqc replied:
Good point. I used to need this but with the current implementation of the socket connections I do not. Removed!

kb1lqc committed:
Removed development code/comments

Removed the #dev comment as well as manual setting of sock to a string.
Instead I changed conn=False to force it to always initialize inside the
loop.

kb1lqc committed:
Removed extra line under if statement

kb1lqc committed:
Replaced return status with reture sendAPRSPacket

Removing unnecessary use of status variable by directly returning output
of sendAPRSPacket.

kb1lqc committed:
Removed useless break statement after return

kb1lqc committed:
Cleaned up commented code and sock initialization

kb1lqc committed:
Removed passing of sock variable to threads

aprs_worker creates it's own socket connection so it's unnecessary to pass
any reference to a socket file descriptor when starting up the thread!

reillyeon commented on faraday/aprs.py:
```diff
@@ -693,14 +699,16 @@ def connectAPRSIS():
-                break
 
-            sleep(10)  # Try to reconnect every 10 seconds
+            sleep(2)  # Try to reconnect every 10 seconds
```
Update the comment?

kb1lqc replied:
Updated

reillyeon commented on faraday/aprs.py:
```diff
@@ -268,11 +302,37 @@ def sendPositions(stations, socket):
+
+        # add telemetry to comments
+        comment = comment + b91Tlm
+        altComment = altComment + b91Tlm
```
comment += b91Tlm
    altComment += b91Tlm

`a += b` is more efficient than `a = a + b`.

kb1lqc replied:
Updated

kb1lqc committed:
Updated reconnect comment and BASE91 comment add

Per @reillyeon a += b is more efficient than a = a + b. Therefore I
changed it.

kb1lqc commented:
@reillyeon if you're going to help pull in the code that is effectively PR #269 then I should bring over the changelong here too. I appreciate the help on these if you are doing both!

reillyeon commented:
PR #269 plus the changes here look good to me.

kb1lqc commented:
Learn something new everyday... By pulling in #271 github automatically merged #269 since it was based on the same branch. This is awesome and in retrospect obvious! This means that as we progress through a big task we can commit mini PR's on it and still have one big PR at the end. If one PR that is a combination of a few smaller PR's in merged then we no longer have to worry about those. I like that!

@kb1lqd for awareness.

kb1lqd commented:
Whoa thanks all and now I learned something new too, very cool PR trait!
